src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,201 @@
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERT, _certificateFile);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERTTYPE, CertificateFileType);
+                easy.SetCurlOption(CURLoption.CURLOPT_KEYPASSWD, password);
+                VerboseTrace(_certificateFile + " created");
+            }
+
+            internal void SetClientCertificateOptions(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.Callback ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.CommonName ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.FileExport);
+
+                _certificateFile = null;
+                switch (s_clientCertificatePolicy)
+                {
+                    case ClientCertificatePolicy.Callback:
+                        {
+                            SetCallbacks(easy);
+                            break;
+                        }
+                    case ClientCertificatePolicy.CommonName :
+                        {
+                            SetClientCertCommonName(easy);
+                            break;
+                        }
+                    default:
+                        ExportToFile(easy);
+                        break;
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_certificateFile == null)
+                {
+                    return;
+                }
+                try
+                {
+                    File.Delete(_certificateFile);
+                }
+                catch (IOException e)
+                {
+                    VerboseTrace("Deleting temporary file:" + e);
+                }
+            }
+
+            internal static ClientCertProvider Create(ClientCertificateOption option, string uriScheme)
+            {
+                if (option == ClientCertificateOption.Manual || !string.Equals(uriScheme, UriSchemeHttps))
+                {
+                    return null;
+                }
+
+                using (var myStore = new X509Store())
+                {
+                    myStore.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection clientCerts = myStore.Certificates.FindTlsClientCertificates();
+                    if (clientCerts.Count == 0)
+                    {
+                        VerboseTrace("0 Client Certs");
+                        return null; 
+                    }
+                    VerboseTrace("Client Certs found " + clientCerts.Count);
+
+                    // NOTE: This is a DEVIATION FROM WinHttpHandler
+                    // Unlike WinHttpHandler, we are going to use only one client certificate.
+                    // libcurl, libssl and even X509Certificate2Collection.Export all are able to handle single client certificate.
+                    // Picking the first available cert
```

Comment:
I am not sure, can @vijaykota  or @bartonjs please comment?

To me the chances seem ```1/n``` (where n = ```clientCerts.Count```). The cert we chose may not be chained the way the server likes etc. 

It seems, when libssl Interop API comes around, we can probably create a callback offered by libssl  (``` SSL_CTX_get_client_cert_cb ```).

For openssl and darwinssl,  if we pass a pfx containing more than 1 cert (after  fixing X509Certificate2Collection.Export ) they still pick the first. gnutls pfx is broken anyway. 

So only other alternative seems to retry with a different client cert if we get a 35 as error code (a blanket ssl error) and see the error message.

At a broader level: If  #3185 gets fixed, may be we can provide a public Property which provides an API for certficate